### PR TITLE
Implement Slippage for Deposit and Sponsor

### PR DESCRIPTION
**Pull request by 0xrin1**
- Implement slippage for deposit and sponsor transactions so that CurveSWapper does not use hardcoded slippage
- Adjust tests to pass slippage

### CODE REVIEW

File: test/Vault.spec.ts [hunk 1]
```diff
@@ -476,7 +476,7 @@ describe('Vault', () => {
       await addUnderlyingBalance(owner, '500');
       await vault
         .connect(owner)
-        .sponsor(underlying.address, parseUnits('500'), TWO_WEEKS);
+        .sponsor(underlying.address, parseUnits('500'), TWO_WEEKS, BigNumber.from(5));
```
Comment (brkomir) [inline test/Vault.spec.ts:479]: maybe introduce a constant for repeating `BigNumber.from(5)`?

File: contracts/test/TestCurveSwapper.sol [hunk 2]
```diff
@@ -23,12 +23,12 @@ contract TestCurveSwapper is CurveSwapper {
         return _underlying;
     }
 
-    function test_swapIntoUnderlying(address _token, uint256 _amount) external {
-        _swapIntoUnderlying(_token, _amount);
+    function test_swapIntoUnderlying(address _token, uint256 _amount, uint256 slippage) external {
```
Comment (brkomir) [inline contracts/test/TestCurveSwapper.sol:26]: missing underscore in param name

File: contracts/test/TestCurveSwapper.sol [hunk 3]
```diff
@@ -23,12 +23,12 @@ contract TestCurveSwapper is CurveSwapper {
...
+        _swapIntoUnderlying(_token, _amount, slippage);
     }
 
-    function test_swapFromUnderlying(address _token, uint256 _amount) external {
-        _swapFromUnderlying(_token, _amount);
+    function test_swapFromUnderlying(address _token, uint256 _amount, uint256 slippage) external {
+        _swapFromUnderlying(_token, _amount, slippage);
     }
```
Comment (brkomir) [inline contracts/test/TestCurveSwapper.sol:32]: maybe I'm missing something but I'm not certain what's the purpose of these tests, what are they asserting?

**Comment by gabrielpoca** [comment]
@0xrin1 please rebase from master to clear up the formatting issues

File: contracts/Vault.sol [hunk 4]
```diff
@@ -1,24 +1,24 @@
...
+import { Counters } from "@openzeppelin/contracts/utils/Counters.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+import { IVault } from "./vault/IVault.sol";
+import { IVaultSponsoring } from "./vault/IVaultSponsoring.sol";
+import { IVaultSettings } from "./vault/IVaultSettings.sol";
+import { CurveSwapper } from "./vault/CurveSwapper.sol";
+import { PercentMath } from "./lib/PercentMath.sol";
+import { IStrategy } from "./strategy/IStrategy.sol";
+import { CustomErrors } from "./interfaces/CustomErrors.sol";
```
[the remark this answers is not in the document]
Comment (0xrin1) [reply contracts/Vault.sol]: fixed

my editor was not picking up local prettier config

File: contracts/test/TestCurveSwapper.sol (hunk 3 above)
Comment (0xrin1) [reply contracts/test/TestCurveSwapper.sol]: not entirely sure tbh but IMO that problem is OOS & better addressed in another PR

File: contracts/Vault.sol [hunk 5]
```diff
@@ -469,7 +470,8 @@ contract Vault is
     function sponsor(
         address _inputToken,
         uint256 _amount,
-        uint256 _lockDuration
+        uint256 _lockDuration,
+        uint256 slippage
```
Comment (brkomir) [inline contracts/Vault.sol:474]: nitpicking, but missing an underscore here on the param name

**Comment by 0xrin1** [comment]
I think I've resolved all the formatting issues now

**Review by brkomir — APPROVED** [review]
LGTM, even though there are some more underscores missing in parameter names :smile:

File: contracts/vault/CurveSwapper.sol [hunk 6]
```diff
@@ -77,10 +74,11 @@ abstract contract CurveSwapper {
...
     /// @param _amount The amount of underlying we want to swap
     /// TODO missing slippage checks
-    function _swapIntoUnderlying(address _token, uint256 _amount)
-        internal
-        returns (uint256 amount)
-    {
+    function _swapIntoUnderlying(
+        address _token,
+        uint256 _amount,
+        uint256 slippage
```
Comment (gabrielpoca) [inline contracts/vault/CurveSwapper.sol:80]: missing _

File: contracts/vault/CurveSwapper.sol [hunk 7]
```diff
@@ -115,10 +114,11 @@ abstract contract CurveSwapper {
...
     /// @param _amount The amount of underlying we want to swap
     /// TODO missing slippage checks
-    function _swapFromUnderlying(address _token, uint256 _amount)
-        internal
-        returns (uint256 amount)
-    {
+    function _swapFromUnderlying(
+        address _token,
+        uint256 _amount,
+        uint256 slippage
```
Comment (gabrielpoca) [inline contracts/vault/CurveSwapper.sol:120]: missing _

File: contracts/vault/IVaultSponsoring.sol [hunk 8]
```diff
@@ -35,7 +35,8 @@ interface IVaultSponsoring {
     function sponsor(
         address _inputToken,
         uint256 _amount,
-        uint256 _lockedUntil
+        uint256 _lockedUntil,
+        uint256 slippage
```
Comment (gabrielpoca) [inline contracts/vault/IVaultSponsoring.sol:39]: missing _

**Comment by 0xrin1** [comment]
added the remaining underscores

**Review by gabrielpoca — APPROVED** [review]

**Pull request merged**